File: Principal Components Analysis.R
#Principal Components Analysis

#read the wine.csv file to a new variable named mydata
mydata <- read.csv("C:\\Users\\yhari\\OneDrive\\Documents\\4. Github\\wine.csv")


#remove the type  column and consider other columns
data <- mydata[-1]
attach(data)


#normalize the data using scale function
normalized <- scale(data)
 
#implementing Principal components analysis on normalized data  
#and store those in a variable call pcawine
pcawine <- princomp(normalized, cor=TRUE)
pcawine

#summary statistics of the normalized data/ probability of variance
summary(pcawine)

#weights on each variables
loadings (pcawine)

#graph showing the importance of PCA
#Comp.1 having the highest importance(highest variance: )
windows()
plot(pcawine)

#PCA scores which represents the whole data
pcawine$scores

# using column bind function, bind the PCA scores to data
#considering only top 6 PCA's 
newdata <- cbind(data, pcawine$scores[, 1:6])

norm_wine <- scale(pcawine$score)
newdata2 <- cbind(norm_wine[,1:6])

#distance matrix and use the "euclidean"method
dista <- dist(newdata2,method="euclidean")

#hierarchical clustering using "complete"method
clus <- hclust(dista, method="complete")
#display dendrogram
plot(clus)

#set the dendrogram hanging level as -3
plot(clus, hang=-3)
#set the numbers of clustering as 7
group <- cutree(clus, k=7)
group

#use the red color borders to divide those 7 culsters
rect.hclust(clus, k=7, border="red")

clusno <- as.data.frame(group)
clusno

#convert into a dataframe
final <- data.frame(clusno,newdata)



data$cluster <- clusno$group
data_clus_1 <- final[final$group ==3,]

#summary statistics of data
summary(data)

#build the model m1 using linear regression function lm
m1 <- lm(Alcohol ~ Malic+Ash+Alcalinity+Magnesium+Phenols+Flavanoids+Nonflavanoids+Proanthocyanins+Color+Hue+Dilution+Proline, data=data_clus_1)
summary(m1)

m2<- lm(Alcohol~Comp.1+Comp.2+Comp.3+Comp.4+Comp.5+Comp.6, data =data_clus_1)
summary(m2)
cor(data)

File: Principal Components Analysis.py
#!/usr/bin/env python
# coding: utf-8

# # PCA (Principal Components Analysis)
# ## wine.cvs

# In[16]:


#importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn.decomposition as sk
from sklearn.decomposition import PCA
from sklearn.preprocessing import scale  


# In[17]:


#imporing the dataset
wine = pd.read_csv("C:\\Users\\yhari\\OneDrive\\Documents\\1. Data Analyst Training\\CSV files\\wine.csv")


# In[18]:


#describe the dataset
wine.describe()


# In[19]:


#view top 5 records
wine.head()


# In[20]:


#Consider only the numerical data
wine.data = wine.iloc[:,1:]


# In[21]:


wine.data.head()


# In[22]:


#Standardizing;Normalizing the numerical data
wine_norm = scale(wine.data)


# In[23]:


#look for the type of data, array or dataframe
type(wine_norm)


# In[24]:


#convert ndarray into a dataframe
wine1 = pd.DataFrame(wine_norm)


# In[25]:


wine1.head(3)


# In[26]:


#creating PCA models
x = PCA()
m1 = PCA().fit_transform(wine1)
m2 = x.fit_transform(wine_norm)


# In[27]:


#shape of the PCA model
m1


# In[28]:


#The amount of varience that each PCA explain is
var = x.explained_variance_ratio_
var


# In[29]:


#components of PCA
x.components_


# In[30]:


#conver the ndarray to a dataframe
wts = pd.DataFrame(x.components_)
wts


# In[31]:


#cumilative varience
cumvar = np.cumsum(np.round(var, decimals = 4)*100)
cumvar


# In[32]:


newdata = pd.DataFrame(cumvar)


# In[33]:


newdata.head()


# In[34]:


#convert the m1 model's ndarray to a dataframe and assign 
#to a new variable call new_wine 

new_wine = pd.DataFrame(m1)
new_wine


# In[35]:


#variance plot for PCA components obtained 
plt.plot(cumvar, color='red')
plt.show()


# In[ ]:




